Source/DafnyCore/ProofDependencyWarnings.cs

```diff
@@ -42,15 +44,98 @@ public static void WarnAboutSuspiciousDependenciesForImplementation(DafnyOptions
     if (results.Any(r => r.Outcome != SolverOutcome.Valid)) {
       return;
     }
+    var unusedFunctions = UnusedFunctions(dafnyOptions, depManager, name, results.SelectMany(r => r.CoveredElements), results.Select(DafnyConsolePrinter.DistillVCResult).SelectMany(r => r.AvailableAxioms));
+    WarnAboutSuspiciousDependencies(dafnyOptions, reporter, depManager, name, results.Select(DafnyConsolePrinter.DistillVCResult).ToList(), unusedFunctions);
+  }
+
+  private static List<Function> UnusedFunctions(DafnyOptions dafnyOptions, ProofDependencyManager depManager, string name, IEnumerable<TrackedNodeComponent> coveredElements, IEnumerable<Axiom> axioms) {
+    if (!(dafnyOptions.Get(CommonOptionBag.SuggestProofRefactoring) && depManager.idsByMemberName[name].Decl is Method)) {
+      return new List<Function>();
+    }
+
+    var unusedFunctions = new List<Function>();
+    if (depManager.idsByMemberName[name].Decl is not Method m) {
+      return unusedFunctions;
+    }
+
+    var referencedFunctions = ReferencedFunctions(m);
+
+    var hiddenFunctions = HiddenFunctions(referencedFunctions);
+
+    var usedFunctions = coveredElements.Select(depManager.GetFullIdDependency).OfType<FunctionDefinitionDependency>()
+      .Select(dep => dep.function).Deduplicate((a, b) => a.Equals(b));
+
+    unusedFunctions = referencedFunctions.Except(hiddenFunctions).Except(usedFunctions).ToList();
 
-    var coveredElements = results.SelectMany(r => r.CoveredElements);
+    HashSet<Function> HiddenFunctions(HashSet<Function> functions) {
```

Comment:
Please move the local function so it sits after the return statement. It's nicer to read statements as a single sequence, instead of one that is interrupted by a local function definition.